src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs

```diff
@@ -386,19 +386,30 @@ private async Task DequeueBlocksContinuouslyAsync()
                 {
                     if (this.batch.Count != 0)
                     {
-                        await this.SaveBatchAsync().ConfigureAwait(false);
-
-                        lock (this.blocksCacheLock)
+                        // Ensure we can retry if an exception is thrown by SaveBatchAsync.
+                        try
                         {
-                            foreach (ChainedHeaderBlock chainedHeaderBlock in this.batch)
+                            await this.SaveBatchAsync().ConfigureAwait(false);
+
+                            // If an error occurred during SaveBatchAsync then this code
+                            // which clears the batch will not execute.
+                            lock (this.blocksCacheLock)
                             {
-                                this.pendingBlocksCache.Remove(chainedHeaderBlock.ChainedHeader.HashBlock);
+                                foreach (ChainedHeaderBlock chainedHeaderBlock in this.batch)
+                                {
+                                    this.pendingBlocksCache.Remove(chainedHeaderBlock.ChainedHeader.HashBlock);
+                                }
+
+                                this.batch.Clear();
                             }
 
-                            this.batch.Clear();
+                            this.currentBatchSizeBytes = 0;
+                        }
+                        catch (Exception err)
```

Comment:
Catch particular exception; 
Add `<Throws` section to SaveBatchAsync with the exception being throws

Reply:
ALL exception types have to be handled. If we miss an exception type the asynchronous loop will exit permanently.

Reply:
Added exception documentation to `SaveBatchAsync` and referenced methods.

Reply:
I disagree we should not catch and continue as usual especially not in such sensitive code, in this case it will mean gap between blocks in store.

We should catch the exception and it should somehow stop the node, unless its this is not a recoverable exception.

Reply:
@dangershony, due to the blocks not being removed from the batch I don't expect any gaps - i.e. this should be an elegant way of allowing the blocks to be `retried` on the next `SaveBatchAsync` - as per the comment on line 394. 

Am I missing something?